### PR TITLE
Disallow rename for elements that are defined in the standard TypeScript...

### PR DESCRIPTION
**Pull request by jramsay**
Disallow rename for elements that are defined in the standard TypeScript Library

### CODE REVIEW

File: src/services/services.ts [hunk 1]
```diff
@@ -5450,22 +5450,39 @@ module ts {
...
+                                createTextSpan(node.getStart(), node.getWidth()));
+                        }
                     }
                 }
             }
 
             return getRenameInfoError(getLocaleSpecificMessage(Diagnostics.You_cannot_rename_this_element.key));
 
+            function endsWith(string: string, value: string): boolean {
+                return string.substring(string.length - value.length, string.length) === value;
```
Comment (CyrusNajmabadi) [inline src/services/services.ts]: any concern about value.length being greater than string.length?  maybe check the lengths first and immediately return false if 'value' couldn't possibly be the end of 'string'.

Comment (CyrusNajmabadi) [inline src/services/services.ts]: Also, you can do this without allocations like so: string.indexOf(value, string.length - value.length) >= 0

Comment (JsonFreeman) [inline src/services/services.ts]: Yes, I would prefer to do this without substring. You can do:
```ts
string.lastIndexOf(value) + value.length === string.length
```

File: src/services/services.ts [hunk 2]
```diff
@@ -5450,22 +5450,39 @@ module ts {
...
-                if (symbol && symbol.getDeclarations() && symbol.getDeclarations().length > 0) {
-                    var kind = getSymbolKind(symbol, typeInfoResolver, node);
-                    if (kind) {
-                        return getRenameInfo(symbol.name, typeInfoResolver.getFullyQualifiedName(symbol), kind,
-                            getSymbolModifiers(symbol),
-                            createTextSpan(node.getStart(), node.getWidth()));
+                if (symbol) {
+                    var declarations = symbol.getDeclarations();
+                    if (declarations && declarations.length > 0) {
+
```
Comment (JsonFreeman) [inline src/services/services.ts]: Remove this newline

File: src/services/services.ts [hunk 3]
```diff
@@ -5450,22 +5450,39 @@ module ts {
...
+                        // Disallow rename for elements that are defined in the standard TypeScript library.
+                        var defaultLibFile = getDefaultLibFileName(host.getCompilationSettings());
+                        for (var i = 0; i < declarations.length; i++) {
+                            var sourceFile = declarations[i].getSourceFile();
+                            if (sourceFile && endsWith(sourceFile.fileName, defaultLibFile)) {
+                                return getRenameInfoError(getLocaleSpecificMessage(Diagnostics.You_cannot_rename_elements_that_are_defined_in_the_standard_TypeScript_library.key));
+                            }
+            	        }
```
Comment (JsonFreeman) [inline src/services/services.ts]: Fix indentation of this brace

**Comment by JsonFreeman** [comment]
:+1:

File: src/services/services.ts (hunk 1 above)
Comment (jramsay) [inline src/services/services.ts]: Thanks - done

File: src/services/services.ts (hunk 3 above)
Comment (jramsay) [inline src/services/services.ts]: Fixed

File: src/services/services.ts (hunk 2 above)
Comment (jramsay) [inline src/services/services.ts]: done

**Comment by CyrusNajmabadi** [comment]
:+1:

**Pull request merged**